lib/systems/damage_system.cpp

```diff
@@ -0,0 +1,43 @@
+#include "systems/damage_system.hpp"
+#include "components/health_component.hpp"
+
+DamageSystem::DamageSystem(std::shared_ptr<CollisionDetector> cd, std::shared_ptr<EntityManager> em, std::shared_ptr<EntityFactory> ef)
+    : BeastSystem(ef, em), collision_detector(cd) {}
+
+void DamageSystem::update(double) {
+    auto entities_with_damage_component = entityManager->getEntitiesByComponent<DamageComponent>();
+    
+    for (auto& [entity_id, damage_comp]: *entities_with_damage_component) {
+        {
+            auto collision = collision_detector->spaceLeft(entity_id, Axis::X, Direction::NEGATIVE);
+            if (collision.space_left >= 0 && collision.object_id && !collision.is_trigger)
+                this->collide(entity_id, damage_comp, collision);
+        }
+        {
+            auto collision = collision_detector->spaceLeft(entity_id, Axis::X, Direction::POSITIVE);
+            if (collision.space_left <= 0 && collision.object_id && !collision.is_trigger)
+                this->collide(entity_id, damage_comp, collision);
+        }
+        {
+            auto collision = collision_detector->spaceLeft(entity_id, Axis::Y, Direction::NEGATIVE);
+            if (collision.space_left >= 0 && collision.object_id && !collision.is_trigger)
+                this->collide(entity_id, damage_comp, collision);
+        }
+        {
+            auto collision = collision_detector->spaceLeft(entity_id, Axis::Y, Direction::POSITIVE);
+            if (collision.space_left <= 0 && collision.object_id && !collision.is_trigger)
+                this->collide(entity_id, damage_comp, collision);
+        }
+    }
+}
+
+void DamageSystem::collide(int entity_id, DamageComponent* damage_comp, CollisionReturnValues collision) {
```

Comment:
Entity_id is not used in this function at the moment. Will this change in the future or can it be removed from the argument list?

Reply:
Removed it